Add zero downtime aura migration guide (#112)
neo4j/docs-kafka-connector: modules/ROOT/pages/migration/neo4j-to-aura.adoc

Reviewer: FYI — in PS we've developed a custom connector that achieves the same benefits as the Source ID strategy (no constraints required, supports parallel relationships) with less prep work (no added elementId property on every entity). Probably not something for this doc, but happy to show it to you in case it's useful for comparison or worth productizing.

Author: Let's catch up on this.

[[zero-downtime-aura-migration]]
= Zero downtime migration to Aura

> This guide describes how to migrate a self-managed Neo4j database to Neo4j Aura without downtime, using the {product-name} to keep the target Aura instance in sync with the source database.

Reviewer: I think "zero downtime" is the right title — it's negligible compared to any other migration strategy — but it's worth being precise somewhere in the intro or the Cutover section: reads stay online throughout, and there is a brief write freeze at cutover while the small number of in-flight changes drain before applications are redirected to Aura.

Author: Added in [b5480ad](https://github.com/neo4j/docs-kafka-connector/pull/112/commits/b5480ad682e50258158f32f1b46614ed5fae1eb4).

The approach relies on link:{page-canonical-root}/cdc/current[Change Data Capture] (CDC) to replay every change committed to the source database after the backup was taken, so that no write is lost during the cutover.
Both the source and sink connectors are configured for exactly-once semantics to guarantee that each change is applied to the target exactly once.

[TIP]
====
At a high level, the migration works as follows:

. A backup captures the source database up to a known transaction.
. The source connector starts streaming changes from the exact point the backup ended, into a dedicated topic.
. The backup is restored into a new Aura instance.
. The sink connector applies the buffered changes to Aura, catching it up to the source.

Because changes are buffered in Kafka from the moment the backup is taken, the source database can keep serving writes throughout the migration.
====

== Prerequisites

> * A Neo4j Enterprise 5 source database with link:{page-canonical-root}/cdc/current[Change Data Capture] support.

Reviewer: Worth adding a prerequisite about privileges: `db.cdc.query` requires admin privileges or grants of both `EXECUTE PROCEDURE db.cdc.query` and `EXECUTE BOOSTED PROCEDURE db.cdc.query` (see [CDC → Getting started](https://neo4j.com/docs/cdc/current/get-started/self-managed/)). This applies both to the source connector's database user and to whoever runs the change-identifier lookup query in Step 4.

Author: Added in [b5480ad](https://github.com/neo4j/docs-kafka-connector/pull/112/commits/b5480ad682e50258158f32f1b46614ed5fae1eb4).

* A running Kafka Connect cluster that supports link:https://cwiki.apache.org/confluence/display/KAFKA/KIP-618%3A+Exactly-Once+Support+for+Source+Connectors[KIP-618] (Apache Kafka 3.3.0 and later) with exactly-once source support enabled on the workers.
* Network connectivity from the Kafka Connect cluster to both the source database and the target Aura instance.

[#choose-strategy]
== Choose a sink strategy

The CDC sink connector can apply change events to the target using one of two sub-strategies, and *you must decide which one to use before taking the backup* in <<backup-source-db>>.
The choice determines whether you need to prepare the source data first, because the Source ID strategy requires markers that have to be present in the backup itself.

xref:sink/cdc.adoc#schema[Schema strategy]:: Merges nodes and relationships by the constraints declared on the source database. This preserves the source schema and adds nothing extra to your data, but it can only be used when the schema requirements below are met.

> xref:sink/cdc.adoc#source-id[Source ID strategy]:: Merges nodes and relationships by the source entity's `elementId`, stored as a property and marked with a dedicated label. It has no constraint requirements, but it adds a marker label and property to every node and relationship.

Reviewer: "It has no constraint requirements" is slightly misleading — Step 6 creates a uniqueness constraint on the marker property on the target. Suggest scoping it:

```suggestion
xref:sink/cdc.adoc#source-id[Source ID strategy]:: Merges nodes and relationships by the source entity's `elementId`, stored as a property and marked with a dedicated label. It has no constraint requirements on the source database (a uniqueness constraint on the marker is still created on the target in <<prepare-target-schema>>), but it adds a marker label and property to every node and relationship.
```

Author: Added in [b5480ad](https://github.com/neo4j/docs-kafka-connector/pull/112/commits/b5480ad682e50258158f32f1b46614ed5fae1eb4).

[#schema-strategy-requirements]
=== Schema strategy requirements

You can use the Schema strategy only if *both* of the following hold on the source database:

> * Every node label that participates in the migration has a *node key* constraint (or a combination of *uniqueness + existence* constraints) covering its key properties, so each node can be identified from the change event alone.

Reviewer: "Every node label ... has a node key constraint" reads stricter than it needs to be — secondary labels can ride along without a key. It's the node that needs at least one constrained label:

```suggestion
* Every node that participates in the migration has at least one label with a *node key* constraint (or a combination of *uniqueness + existence* constraints) covering its key properties, so each node can be identified from the change event alone. Secondary labels without constraints are carried along as-is.
```

Author: Added in [b5480ad](https://github.com/neo4j/docs-kafka-connector/pull/112/commits/b5480ad682e50258158f32f1b46614ed5fae1eb4).

* There are no duplicate relationships of the same type between the same pair of nodes, *unless* those relationships have a *relationship key* constraint to tell them apart.

If you can satisfy both requirements, continue with the Schema strategy.
The migration steps that follow apply unchanged, and you skip the Source ID-specific source preparation in <<source-prep>>.

> === When to use Source ID

Reviewer: Reading this end to end, I wonder how strongly we should recommend Source ID at all — stamping a label and property onto every entity is a big ask and hard to guarantee, especially with ongoing ingestion: anything created after the stamping queries finish but before the backup lands in the backup unmarked, and its later change events won't match on the target. Suggest we (a) explicitly position Schema as the preferred strategy with Source ID as the fallback, and (b) strongly recommend creating an existence constraint on the source before the backup so unmarked nodes are caught:

```
CREATE CONSTRAINT source_event_source_id_exists IF NOT EXISTS
FOR (n:SourceEvent) REQUIRE n.sourceId IS NOT NULL
```

(Noting the limitation: this catches labeled nodes missing the property, not nodes created without the label — so guidance that ongoing ingestion should write the markers itself, or that the stamping queries be re-run just before the backup, would close the gap further.)

Author: Added in [b5480ad](https://github.com/neo4j/docs-kafka-connector/pull/112/commits/b5480ad682e50258158f32f1b46614ed5fae1eb4).

> If you *cannot* satisfy the Schema requirements — for example, some labels have no usable key, or duplicate relationships exist without a relationship key — use the Source ID strategy instead.

Reviewer: Same fix as the requirements above — it's nodes, not labels, that lack a usable key:

```suggestion
If you *cannot* satisfy the Schema requirements — for example, some nodes have no label with a usable key, or duplicate relationships exist without a relationship key — use the Source ID strategy instead.
```

Author: Added in [b5480ad](https://github.com/neo4j/docs-kafka-connector/pull/112/commits/b5480ad682e50258158f32f1b46614ed5fae1eb4).

[IMPORTANT]
====
The Source ID strategy identifies each entity by the source `elementId`, stored as a property (default `sourceId`) on a marked node (default label `SourceEvent`).
In a backup-based migration, the nodes and relationships restored from the backup must *already* carry these markers, otherwise change events for pre-existing entities will not match anything on the target and the sink will create duplicates.

> For this reason, you must add the marker label and property to *every node and relationship on the source database before taking the backup*.

Reviewer: Worth a warning here about cost: on large databases, this prep is a full-graph update — a new property on every node and relationship, plus transaction log churn and lock contention if ingestion is ongoing. The single marker index created later on the target is also a full-graph index build. Users with billions of entities should know up front that this step can be prohibitive before they commit to the strategy.

Author: Added in [b5480ad](https://github.com/neo4j/docs-kafka-connector/pull/112/commits/b5480ad682e50258158f32f1b46614ed5fae1eb4).

Add the marker label and property to all nodes:

[source,cypher]
----
MATCH (n)
CALL (n) {
  SET n:SourceEvent, n.sourceId = elementId(n) //<1>
} IN TRANSACTIONS OF 10000 ROWS
----

Add the marker property to all relationships:

[source,cypher]
----
MATCH ()-[r]->()
CALL (r) {
  SET r.sourceId = elementId(r) //<1>
} IN TRANSACTIONS OF 10000 ROWS
----

<1> The label `SourceEvent` and property `sourceId` used here are the defaults.
If you override them, they must match the `neo4j.cdc.source-id.label-name` and `neo4j.cdc.source-id.property-name` settings on the sink connector in <<setup-sink-connector>>.

[NOTE]
Run these statements from a client that allows implicit transactions, such as `cypher-shell`, and adjust the batch size to suit your database.

=====

====

[#enable-cdc]
== Step 2: Enable CDC on the source database

Enable Change Data Capture on the source database before taking the backup, so that all changes committed from this point on are captured and can be replayed onto the target.

> Follow link:{page-canonical-root}/cdc/current/getting-started[Change Data Capture -> Getting Started] to enable the CDC mode on the database.

Reviewer: Should we recommend enabling CDC in `DIFF` mode here? Smaller change-event payloads could improve throughput and the connector's ability to keep up with the source change rate during catch-up. If both sink strategies are fully supported with DIFF events, it seems like the better default for a migration; if not, worth stating which mode to use explicitly.

Author: Added in [b5480ad](https://github.com/neo4j/docs-kafka-connector/pull/112/commits/b5480ad682e50258158f32f1b46614ed5fae1eb4).

Once CDC is active, record the *current change identifier*:

[source,cypher]
----
CALL db.cdc.current() YIELD id
RETURN id
----

Note the returned `id` value: it marks the point in the change stream from which you search for the backup's last transaction in <<setup-source-connector, Step 4>>.
Starting that search from this identifier, rather than from the earliest available change, lets it scan only the changes committed from this point on instead of reading through the entire available transaction log.

[IMPORTANT]
====
Configure the database with an adequate transaction log retention period, set *above the expected backup time*.
The change identifier the source connector resumes from must still be available once the backup completes and source connector starts, so changes committed during the backup must not be purged before the connector can stream them.
See link:https://neo4j.com/docs/cdc/current/get-started/self-managed/#log-retention[Change Data Capture -> Log retention] for more information.
====

[#backup-source-db]
== Step 3: Back up the source database

> Take a backup of the source database and note the *latest transaction ID* included in the backup.

Reviewer: We tell readers to note the latest transaction ID but not how to get it. Suggest showing the command — the HIGHEST TX value from:

```
bin/neo4j-admin backup inspect --show-metadata /path/to/backup/containing/dir/
```

(On Neo4j 5.x the equivalent is `neo4j-admin database backup --inspect-path=<path>` — since the prerequisites target Neo4j 5, we should show the form matching the supported versions, or both.)

Author: Added in [b5480ad](https://github.com/neo4j/docs-kafka-connector/pull/112/commits/b5480ad682e50258158f32f1b46614ed5fae1eb4).

== Step 4: Configure the source connector

The source connector must start streaming from the first change that occurred *after* the last transaction included in the backup.
Streaming from any earlier point would re-apply changes already present in the backup, and starting later would lose changes.

=== Identify the change identifier to resume from

Run the following query against the source database to find the CDC change identifier that corresponds to the last transaction captured in the backup. Substitute:

* `$changeIdAfterCdcEnabled` with the current change identifier you recorded in <<enable-cdc, Step 2>>.
* `$lastTxIdFromBackup` with the latest transaction ID you noted in <<backup-source-db, Step 3>>.

[source,cypher]
----
CALL db.cdc.query($changeIdAfterCdcEnabled) YIELD id, txId, seq
WHERE txId = $lastTxIdFromBackup
ORDER BY seq DESC
RETURN id, txId, seq
LIMIT 1
----

Starting the search from the change identifier recorded right after CDC was enabled — rather than from `db.cdc.earliest()` — keeps the query fast: it scans only the changes committed between enabling CDC and the backup, instead of reading through the entire available transaction log.

> If the query returns a row, note the returned `id` value: this is the change identifier the source connector resumes from.

Reviewer: Suggest adding a reassurance here that the connector resumes from the first change after the provided change identifier (exclusive) — i.e. the change returned by this query is itself already in the backup and will not be re-applied. As written, a careful reader may worry we're replaying the backup's last transaction twice.

Author: Added in [b5480ad](https://github.com/neo4j/docs-kafka-connector/pull/112/commits/b5480ad682e50258158f32f1b46614ed5fae1eb4).

Without it, the target graph could end up missing writes or replaying them twice, silently diverging from the source.

The CDC source connector declares itself exactly-once capable, and from version 5.2.0 takes advantage of link:https://cwiki.apache.org/confluence/display/KAFKA/KIP-618%3A+Exactly-Once+Support+for+Source+Connectors[KIP-618] on workers that support it.
To make this a hard requirement for the migration, the configuration above sets `exactly.once.support` to `required` together with the supporting producer settings, so the connector refuses to start unless exactly-once delivery is actually available.

This requires *exactly-once source support to be enabled on the Kafka Connect workers themselves* (`exactly.once.source.support=enabled` in the worker configuration), which cannot be set per connector.
====

See xref:source/cdc.adoc#_exactly_once_semantics[Source -> Exactly Once Semantics] for more information.

[#provision-aura-db]
== Step 5: Provision the target Aura instance

> Upload the backup taken in <<backup-source-db>> and provision the target Aura database from it.

Reviewer: Suggest a TIP here: this is the last practical opportunity to compact the store (using `neo4j-admin database copy`), since store copy isn't available once you're in Aura. It can be run offline against the restored backup, so no source downtime. Two caveats worth stating if we add it: copy discards indexes/constraints (Step 6 recreates them anyway), and it reassigns internal IDs — which doesn't break either strategy, since Schema matches by keys and Source ID matches by the stored `sourceId` property values, which are copied along.

Author: Added in [b5480ad](https://github.com/neo4j/docs-kafka-connector/pull/112/commits/b5480ad682e50258158f32f1b46614ed5fae1eb4).

For details on importing a backup into Aura, see link:{page-canonical-root}/aura/managing-instances/importing-data/[Aura -> Import data].

[#prepare-target-schema]
== Step 6: Prepare the target schema

=== Strategy-specific schema

[.tabbed-example]
====
[.include-with-Schema]
=====
Make sure that *all constraints present on the source database are also defined on the target Aura instance*.
The Schema strategy merges nodes and relationships using the node key, relationship key, and uniqueness/existence constraints declared in the change events, so the corresponding constraints must exist on the target for changes to be applied correctly.

List the constraints on the source database with:

[source,cypher]
----
SHOW CONSTRAINTS
----

Recreate each of them on the target Aura instance.
=====

[.include-with-Source-ID]
=====

> The Source ID strategy does not rely on the source constraints to identify entities, so they are not required on the target.

Reviewer: True that the source constraints aren't needed for the replay itself — but we should still recommend recreating all source constraints (and indexes) on the target before more data is added, and at the latest before cutover. The replication mirrors a source that satisfied them, so creating them early is safe, and building large indexes after cutover time adds risk.

Author: Added in [b5480ad](https://github.com/neo4j/docs-kafka-connector/pull/112/commits/b5480ad682e50258158f32f1b46614ed5fae1eb4).

[[zero-downtime-aura-migration]]
= Zero downtime migration to Aura

This guide describes how to migrate a self-managed Neo4j database to Neo4j Aura without downtime, using the {product-name} to keep the target Aura instance in sync with the source database.

Throughout the migration, the source database stays online for both reads and writes.

> The only interruption is a short write freeze during <<cutover>>, while the last in-flight changes drain before applications are redirected to Aura; reads are never interrupted.

Reviewer 2: ```suggestion
The only interruption is a short write freeze during <<cutover>>, while the last in-flight changes drain before applications are redirected to Aura; reads are not interrupted.
```

Author: Addressed in ca2102a.

The approach relies on link:{page-canonical-root}/cdc/current[Change Data Capture] (CDC) to replay every change committed to the source database after the backup was taken, so that no write is lost during the cutover.
Both the source and sink connectors are configured for exactly-once semantics to guarantee that each change is applied to the target exactly once.

> [TIP]

Reviewer 2: This is not a tip, so I would recommend to remove the admonition style here.

Author: Addressed in ca2102a.

xref:sink/cdc.adoc#schema[Schema strategy]:: Merges nodes and relationships by the constraints declared on the source database. This preserves the source schema and adds nothing extra to your data, but it can only be used when the schema requirements below are met.
xref:sink/cdc.adoc#source-id[Source ID strategy]:: Merges nodes and relationships by the source entity's `elementId`, stored as a property and marked with a dedicated label. It has no constraint requirements on the source database (a uniqueness constraint on the marker is still created on the target in <<prepare-target-schema>>), but it adds a marker label and property to every node and relationship.

*The Schema strategy is the preferred choice whenever the source schema allows it.*
It adds nothing to your data and preserves the source schema.
Use the Source ID strategy only as a fallback for databases that cannot meet the Schema requirements below: it works without constraints, but marking every entity in the source is a substantial preparation step with the trade-offs called out in <<source-prep, Step 1>>.

[#schema-strategy-requirements]
=== Schema strategy requirements

> You can use the Schema strategy only if *both* of the following hold on the source database:

Reviewer 2: ```suggestion
You can use the Schema strategy only if the source database meet *both* of the following requirements:
```

Author: Addressed in ca2102a.

If you can satisfy both requirements, continue with the Schema strategy.
The migration steps that follow apply unchanged, and you skip the Source ID-specific source preparation in <<source-prep>>.

=== When to use Source ID

If you *cannot* satisfy the Schema requirements — for example, some nodes have no label with a usable key, or duplicate relationships exist without a relationship key — use the Source ID strategy instead.

[IMPORTANT]
====
The Source ID strategy identifies each entity by the source `elementId`, stored as a property (default `sourceId`) on a marked node (default label `SourceEvent`).
In a backup-based migration, the nodes and relationships restored from the backup must *already* carry these markers, otherwise change events for pre-existing entities will not match anything on the target and the sink will create duplicates.

For this reason, you must add the marker label and property to *every node and relationship on the source database before taking the backup*.
This is covered in <<source-prep>>.
====

> [WARNING]

Reviewer 2: Three admonition boxes in a row is bad UX and should be avoided. Is it possible to remove at least one? I would remove the [NOTE] and just keep that as running text. Another way to bring extra attention to something is by making certain words *bold*.

Author: Removed two, and kept the `WARNING` one. Addressed in ca2102a.

  }
}
----

<1> Use `USER_PROVIDED` so the connector starts from a specific change identifier rather than `EARLIEST` or `NOW`.
<2> The change identifier (`id`) returned by the query above.
<3> Capture all node and relationship changes into the dedicated `migration` topic. Adjust the patterns if you only need to migrate a subset of the graph. See xref:source/cdc.adoc#patterns[Patterns] for details.
<4> Set the key strategy to `SKIP` so change events are produced without a record key, reducing the overall message size. Combined with the single-partition migration topic, this keeps all change events on the one partition and preserves their order.
<5> Require exactly-once support, so the connector fails to start rather than silently falling back to at-least-once delivery if the worker is not configured for it.
<6> Define transactions on each poll, so every batch of change events is committed to the topic as a single Kafka transaction.
<7> Require acknowledgement from all in-sync replicas before a write is considered successful, so no published change event can be lost.
<8> Enable the idempotent producer to prevent duplicate change events from producer retries.

> [IMPORTANT]

Reviewer 2: Both of these admonitions need to be removed. The text inside them should use a different strategy for emphasis. An admonition (whether NOTE, WARNING, TIP, or IMPORTANT or anything else really) should be limited to a few lines and the number of admonitions on a page should be few. They lose their purpose otherwise.

Author: Addressed in ca2102a.

Without it, the target graph could end up missing writes or replaying them twice, silently diverging from the source.

The CDC source connector declares itself exactly-once capable, and from version 5.2.0 takes advantage of link:https://cwiki.apache.org/confluence/display/KAFKA/KIP-618%3A+Exactly-Once+Support+for+Source+Connectors[KIP-618] on workers that support it.
To make this a hard requirement for the migration, the configuration above sets `exactly.once.support` to `required` together with the supporting producer settings, so the connector refuses to start unless exactly-once delivery is actually available.

This requires *exactly-once source support to be enabled on the Kafka Connect workers themselves* (`exactly.once.source.support=enabled` in the worker configuration), which cannot be set per connector.
====

See xref:source/cdc.adoc#_exactly_once_semantics[Source -> Exactly Once Semantics] for more information.

[#provision-aura-db]
== Step 5: Provision the target Aura instance

Upload the backup taken in <<backup-source-db>> and provision the target Aura database from it.

> [TIP]

Reviewer 2: As above, this is way too long for a TIP.

Author: Removed the TIP, but what do you think would be best way to mark this as an optional thing?

Reviewer 2: You could rephrase and say something like:

_If you want to compact the store <insert reason why you'd want to do that here maybe> using link:{page-canonical-root}/operations-manual/current/tools/neo4j-admin/copy-database/[`neo4j-admin database copy`] before you upload the database to Aura, you can do so at this point.
This step is *optional*, but once the database is in Aura, this command is not supported._

Reviewer 2: optional should be bold but it doesn't render here..

====

For details on importing a backup into Aura, see link:{page-canonical-root}/aura/managing-instances/importing-data/[Aura -> Import data].

[#prepare-target-schema]
== Step 6: Prepare the target schema

=== Strategy-specific schema

[.tabbed-example]
====
[.include-with-Schema]
=====
No strategy specific schema is required for the Schema strategy.
=====

[.include-with-Source-ID]
=====
Create a uniqueness constraint on the marker property so that each source entity maps to exactly one target node and lookups by the marker are backed by an index:

[source,cypher]
----
CREATE CONSTRAINT source_event_source_id IF NOT EXISTS
FOR (n:SourceEvent) REQUIRE n.sourceId IS UNIQUE
----

Use the label and property you configured for the strategy if you overrode the defaults.

=====
====

=== Ensure constraints and indexes are present

Regardless of the strategy, list the constraints and indexes on the source database with:

[source,cypher]
----
SHOW CONSTRAINTS
----

[source,cypher]
----
SHOW INDEXES
----

> and make sure each of them are also available on the target Aura instance.

Reviewer 2: ```suggestion
Make sure each of them are also available on the target Aura instance.
```

Author: Addressed in ca2102a.
